tests/integration/components/bs-button-test.js

```diff
@@ -104,52 +112,163 @@ module('Integration | Component | bs-button', function(hooks) {
   });
 
   test('button text is changed according to button state', async function(assert) {
+    this.set('clickAction', () => {
+      return new EmberPromise((resolve) => {
+        later(() => {
+          resolve();
+        }, 10);
+      });
+    });
 
-    await render(hbs`{{bs-button defaultText="text1" loadingText="text2" textState=textState}}`);
-    assert.dom('button').hasText('text1');
+    await render(
+      hbs`{{bs-button
+      defaultText="default text"
+      pendingText="text for pending state"
+      resolvedText="text for resolved state"
+      rejectedText="text for rejected state"
+      onClick=clickAction
+    }}`);
+    assert.dom('button').hasText('default text');
+
+    click('button');
+    await nextRunloop();
+    assert.dom('button').hasText('text for pending state');
+
+    await settled();
+    assert.dom('button').hasText('text for resolved state');
+
+    this.set('clickAction', () => {
+      return new EmberPromise((resolve, reject) => {
+        later(() => {
+          reject();
+        }, 10);
+      });
+    });
+    click('button');
+    await nextRunloop();
+    assert.dom('button').hasText('text for pending state');
 
-    run(() => this.set('textState', 'loading'));
+    await settled();
+    assert.dom('button').hasText('text for rejected state');
+  });
+
+  test('setting reset to true resets button text', async function(assert) {
+    this.set('clickAction', () => {
+      return resolve();
+    });
 
-    assert.dom('button').hasText('text2');
+    await render(hbs`
+      {{bs-button defaultText="default text" resolvedText="text for resolved state" reset=reset onClick=clickAction}}
+    `);
+    assert.dom('button').hasText('default text');
 
-    run(() => this.set('textState', 'default'));
+    await click('button');
+    assert.dom('button').hasText('text for resolved state');
 
-    assert.dom('button').hasText('text1');
+    run(() => this.set('reset', true));
+    assert.dom('button').hasText('default text');
   });
 
-  test('setting reset to true resets button state', async function(assert) {
-    await render(hbs`{{bs-button defaultText="text1" loadingText="text2" textState=textState reset=reset}}`);
-    run(() => this.set('textState', 'loading'));
+  test('button state is yielded', async function (assert) {
+    this.set('clickAction', () => {
+      return new EmberPromise((resolve) => {
+        later(() => {
+          resolve();
+        }, 10);
+      });
+    });
+    this.set('reset', false);
 
-    assert.dom('button').hasText('text2');
+    await render(hbs`{{#bs-button onClick=clickAction reset=reset as |button|}}{{button.state}}{{/bs-button}}`);
+    assert.dom('button').hasText('default');
```

Comment:
By the way: I don't like the naming `default` for a `state`  if the button wasn't clicked yet. It was fitting for `defaultText` but as a state it doesn't feel intuitive. Do you have a better naming before we expose that one as public API? As an alternative we may just that `state` to `null` if but wasn't clicked yet? We might or might not also rename `defaultText` to `text` in that case.